### PR TITLE
[PAPP-34684] Splunk: Feature - Added 'splunk_job_timeout' parameter to asset config

### PR DESCRIPTION
**Pull request by bbielinski-splunk**
Please ensure your pull request (PR) adheres to the following guidelines:

- Please refer to our contributing documentation for any questions on submitting a pull request, link: [Contribution Guide](https://github.com/splunk-soar-connectors/.github/blob/main/.github/CONTRIBUTING.md)

#### Please check if your PR fulfills the following requirements:
- [x] Testing of all the changes has been performed (for bug fixes / features)
- [x] The manual_readme_content.md has been reviewed and added / updated if needed (for bug fixes / features)
- [x] Use the following format for the PR description: `<App Name>: <PR Type> - <PR Description>`
- [x] Provide release notes as part of the PR submission which describe high level points about the changes for the upcoming GA release.
- [x] Verify all checks are passing.
- [x] Do NOT use the `next` branch of the forked repo. Create separate feature branch for raising the PR.
- [x] Do NOT submit updates to dependencies unless it fixes an issue.
#### Please check the type of change your PR introduces:
- [ ] New App
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation
- [ ] Other (please describe):

## Security Considerations (REQUIRED)
- [ ] If you are exposing any endpoints using a [REST handler](https://docs.splunk.com/Documentation/SOAR/current/DevelopApps/RESTHandlers),
  please document them in the `manual_readme_content.md`.
- [ ] If this is a new connector or you are adding new actions
    - [ ] Please document in the `manual_readme_content.md` all methods (eg, OAuth) used to authenticate
      with the service that the connector is integrating with.
    - [ ] If any actions are unable to run on SOAR Cloud, please document this in the `manual_readme_content.md`.
- [ ] Are you introducing any new cryptography modules? If yes, please elaborate their purpose:
- [ ] Are you are accessing the file system? If yes, please verify that you are only accessing paths returned through
the [Vault](https://docs.splunk.com/Documentation/SOARonprem/5.2.1/DevelopApps/AppDevAPIRef#Vault) API.
- [ ] Are you are marking code to be ignored by Semgrep with [`nosemgrep`](https://semgrep.dev/docs/ignoring-files-folders-code/#ignoring-code-through-nosemgrep)?
If yes, please provide justification in an additional comment next to the ignored code.

## Release Notes (REQUIRED)
Changelog:
- Added 'splunk_job_timeout' parameter to asset config
- Bumped splunk-sdk to 2.0.1 (in 2.x, a lot of legacy Python2 code was removed and newer http libraries are used, which should make the sdk more robust)

## What is the current behavior? (OPTIONAL)
The splunk connector doesn't track the amount of time it waits for the splunk job to finish so if job is stuck in "QUEUED" state (see related PAPP), the action will run indefinitely.

## What is the new behavior? (OPTIONAL)
The splunk connector will track the amount of time it waits for the splunk job to finish and will time out this job if it exceeds set limit

## Other information (OPTIONAL)
See screenshots

## Pay close attention to (OPTIONAL)
n/a 

## Screenshots (if relevant)

New parameter "The duration in seconds to wait before a scheduled Splunk job times out" in asset settings:

Screenshot 2024-09-11 at 11 08 14

Error message, when the job has timed out:
Screenshot 2024-09-10 at 13 51 32

---
Thanks for contributing!

### CODE REVIEW

File: splunk_connector.py [hunk 1]
```diff
@@ -1341,6 +1330,26 @@ def _run_query(self, search_query, action_result, attach_result=False, kwargs_cr
...
+                #  could be stuck in permanent "QUEUED" state after the Splunk
+                #  stack has crashed.
+                max_waiting_time: float = time.time() + self.splunk_job_timeout
+                while not job.is_ready():
+                    if time.time() > max_waiting_time:
+                        return action_result.set_status(phantom.APP_ERROR, consts.SPLUNK_ERR_SPLUNK_JOB_HAS_TIMED_OUT)
+                    time.sleep(self.sleeptime_in_requests)
+                job.refresh()
+                break
+            except Exception as e:
```
Comment (mposluszny-splunk) [inline splunk_connector.py:1346]: can we not generalize this Exception?
or can we debug print if exception happened?

File: splunk_connector.py [hunk 2]
```diff
@@ -1270,7 +1267,7 @@ def _run_query(self, search_query, action_result, attach_result=False, kwargs_cr
         # Create the job
         for attempt_count in range(0, RETRY_LIMIT):
             try:
-                job = self._service.jobs.create(search_query, **kwargs_create)
+                job: splunk_client.Job = self._service.jobs.create(search_query, **kwargs_create)
```
Comment (mposluszny-splunk) [inline splunk_connector.py:1270]: can we do the same generalization as for `_wait_until_splunk_job_results_are_ready` but for `jobs.create`?

File: splunk_connector.py [hunk 3]
```diff
@@ -1341,6 +1330,26 @@ def _run_query(self, search_query, action_result, attach_result=False, kwargs_cr
...
+                #  stack has crashed.
+                max_waiting_time: float = time.time() + self.splunk_job_timeout
+                while not job.is_ready():
+                    if time.time() > max_waiting_time:
+                        return action_result.set_status(phantom.APP_ERROR, consts.SPLUNK_ERR_SPLUNK_JOB_HAS_TIMED_OUT)
+                    time.sleep(self.sleeptime_in_requests)
+                job.refresh()
+                break
+            except Exception as e:
+                if attempt_count == retry_limit - 1:
```
Comment (mposluszny-splunk) [inline splunk_connector.py:1347]: can we make this a bit cleaner? Like for it to be attempt_count == retry_limit (change range appropriately for ex.)

File: splunk_connector.py [hunk 4]
```diff
@@ -1341,6 +1330,26 @@ def _run_query(self, search_query, action_result, attach_result=False, kwargs_cr
         self.debug_print('Done run query')
         return action_result.set_status(phantom.APP_SUCCESS)
 
+    def _wait_until_splunk_job_results_are_ready(self, action_result: phantom.ActionResult, job: splunk_client.Job, retry_limit: int) -> bool:
+        for attempt_count in range(0, retry_limit):
+            try:
```
Comment (mposluszny-splunk) [inline splunk_connector.py:1335]: do we need try: except: on the whole block?

File: splunk_connector.py (hunk 3 above)
Comment (bbielinski-splunk) [reply splunk_connector.py:1347]: Sure, changed

File: splunk_connector.py (hunk 1 above)
Comment (bbielinski-splunk) [reply splunk_connector.py:1346]: Adding less generalize Exception will be troublesome as we don't have any influence over what is raised by the splunk-sdk and when (but every exception should be treated the same way by us for now) so I went with general Exception. 
However, I will add debug print, which should help with tracking the loop with retries

File: splunk_connector.py (hunk 2 above)
Comment (bbielinski-splunk) [reply splunk_connector.py:1270]: Yeah, we can, good idea

**Review by mposluszny-splunk — APPROVED** [review]

**Review by gbanas-splunk — APPROVED** [review]

**Pull request merged**